xarray/core/dataset.py

```diff
@@ -1096,6 +1096,78 @@ def maybe_chunk(name, var, chunks):
                                  for k, v in self.variables.items()])
         return self._replace_vars_and_dims(variables)
 
+    def _validate_indexers(self, indexers):
+        """ Here we make sure
+        + indexer has a valid keys
+        + indexer is in a valid data type
+        + raise an Error for some confusing case.
+        """
+        from .dataarray import DataArray
+
+        invalid = [k for k in indexers if k not in self.dims]
+        if invalid:
+            raise ValueError("dimensions %r do not exist" % invalid)
+
+        # all indexers should be int, slice, np.ndarrays, or Variable
+        indexers_list = []
+        for k, v in iteritems(indexers):
+            if isinstance(v, integer_types + (slice, Variable)):
+                pass
+            elif isinstance(v, DataArray):
+                v = v.variable
+            elif isinstance(v, tuple):
+                v = as_variable(v)
+            elif isinstance(v, Dataset):
+                raise TypeError('cannot use a Dataset as an indexer')
+            else:
+                v = np.asarray(v)
+            indexers_list.append((k, v))
+        return indexers_list
+
+    def _get_indexers_coordinates(self, indexers):
+        """  Extract coordinates from indexers.
+        Returns an OrderedDict mapping from coordinate name to the
+        coordinate variable.
+
+        Coordinates to be extracted and attached should satisfy
+        + Dimension coordinate of the indexers.
+          Non-dimension coordinate of the indexers are not attached.
+        + Only coordinate with a name different from any of sef.variables.
+
+        If self already has the same name coordinate, we raise an ValueError.
+        """
+        from .dataarray import DataArray
+        import warnings
```

Comment:
please at the top level (we only import inside methods to avoid circular import problems)

Reply:
Done